crates/subspace-networking/src/utils.rs

```diff
@@ -95,3 +97,150 @@ pub(crate) fn convert_multiaddresses(addresses: Vec<Multiaddr>) -> Vec<PeerAddre
         })
         .collect()
 }
+
+// Semaphore like implementation that allows both shrinking and expanding
+// the max permits.
+#[derive(Clone, Debug)]
+pub(crate) struct ResizableSemaphore(Arc<SemShared>);
+
+// The permit.
+#[derive(Clone, Debug)]
+pub(crate) struct ResizableSemaphorePermit(Arc<SemShared>);
+
+// The state shared between the semaphore and the outstanding permits.
+#[derive(Debug)]
+struct SemShared {
+    // The tuple holds (current usage, current max capacity)
+    current: Mutex<SemState>,
+
+    // To signal waiters for permits to be available
+    notify: Notify,
+}
+
+// Current state.
+#[derive(Debug)]
+struct SemState {
+    // The current capacity
+    capacity: usize,
+
+    // The current outstanding permits
+    usage: usize,
+}
+
+impl SemShared {
+    fn new(capacity: usize) -> Self {
+        Self {
+            current: Mutex::new(SemState { capacity, usage: 0 }),
+            notify: Notify::new(),
+        }
+    }
+
+    // Allocates a permit if available.
+    // Returns true if a permit was allocated, false otherwise
+    async fn alloc_one(&self) -> bool {
```

Comment:
More like `try_alloc_one` since it doesn't necessarily allocate a permit, moreover result is not `#[must_used]`, so it is possible to just ignore the result thinking permit is allocated.

Of course this is an internal API and it is easy to verify what it does, I'm just sharing the though process for writing code that is misuse-resistant that I follow automatically.

Reply:
We don't have a corresponding blocking version, so `try_alloc_one()` won't be necessary. But it does have `acquire()` and `try_acquire()` as one of them can block

`#[must_used]` needs us to differentiate between functions whose return values should be checked vs not. Since these are internal helpers, think we can leave it as is